File: problems/p0150_evaluate_reverse_polish_notation/solution-0150.java
package p0150_evaluate_reverse_polish_notation;

import java.util.Arrays;
import java.util.Stack;

import static org.junit.jupiter.api.Assertions.assertEquals;
import org.junit.jupiter.api.Test;

class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();
        for (String token : tokens) {
            switch (token) {
                case "+" -> {
                    int rhs = stack.pop();
                    int lhs = stack.pop();
                    stack.push(lhs+rhs);
                }
                case "-" -> {
                    int rhs = stack.pop();
                    int lhs = stack.pop();
                    stack.push(lhs-rhs);
                }
                case "*" -> {
                    int rhs = stack.pop();
                    int lhs = stack.pop();
                    stack.push(lhs*rhs);
                }
                case "/" -> {
                    int rhs = stack.pop();
                    int lhs = stack.pop();
                    stack.push(lhs/rhs);
                }
                default -> {
                    stack.push(Integer.valueOf(token));
                }
            }
        }
        return stack.pop();
    }
}

class TestCase {
    public String[] tokens;
    public int expected;

    public TestCase(String[] tokens, int expected) {
        this.tokens = tokens;
        this.expected = expected;
    }
}

@SuppressWarnings("unused")
class SolutionTest {
    @Test
    void testBwa() {
        Solution solution = new Solution();
        TestCase[] testCases = new TestCase[] {
            new TestCase(new String[] {"4", "5", "*"}, 20),
            new TestCase(new String[] {"3", "2", "5", "+", "*"}, 21),
            new TestCase(new String[] {"3", "6", "-", "14", "*"}, -42),
        };
        for (TestCase testCase : testCases) {
            System.err.println("Input: " + Arrays.toString(testCase.tokens));
            int result = solution.evalRPN(testCase.tokens);
            System.err.println("Expected: " + testCase.expected + " | Result: " + result);

            assertEquals(result, testCase.expected);
        }
    }
}

File: problems/p0150_evaluate_reverse_polish_notation/solution-0150.rs
#![allow(unused)]
pub struct Solution;

macro_rules! operation {
    ($stack: ident, $op:tt) => {
        let rhs = $stack.pop().unwrap();
        let lhs = $stack.pop().unwrap();
        $stack.push(lhs $op rhs)
    };
}

impl Solution {
    pub fn eval_rpn(tokens: Vec<String>) -> i32 {
        let mut stack: Vec<i32> = vec![];
        for token in tokens {
            match token.as_str() {
                "+" => {
                    operation!(stack, +);
                }
                "-" => {
                    operation!(stack, -);
                }
                "*" => {
                    operation!(stack, *);
                }
                "/" => {
                    operation!(stack, /);
                }
                _ => stack.push(token.parse::<i32>().unwrap()),
            }
        }
        stack.pop().unwrap()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    struct TestCase {
        tokens: Vec<String>,
        expected: i32,
    }

    macro_rules! stringify {
        ($vec:expr) => {
            $vec.iter().map(|s| s.to_string()).collect()
        };
    }
    #[test]
    fn test_solution() {
        let cases = vec![
            TestCase {
                tokens: stringify!(vec!["4", "5", "*"]),
                expected: 20,
            },
            TestCase {
                tokens: stringify!(vec!["3", "2", "5", "+", "*"]),
                expected: 21,
            },
            TestCase {
                tokens: stringify!(vec!["3", "6", "-", "14", "*"]),
                expected: -42,
            },
        ];
        for case in cases {
            println!(
                "Testing case with RPN: {:?}, expecting {}",
                case.tokens, case.expected
            );
            let result = Solution::eval_rpn(case.tokens);
            assert_eq!(result, case.expected);
        }
    }
}
